bionic/code_analyzer.py

```diff
@@ -0,0 +1,134 @@
+"""
+This module implements the logic to analyze your code objects and
+returns the result in the form of a byte array.
+"""
+
+import inspect
+import json
+import warnings
+
+from .utils.misc import oneline
+
+
+DEFAULT_NAME = b"zA)T+ndu/Ygd%JH=eux)[F9:<>)zND]}u4x,c49tWH<yd5+#+4Xg{=J<[S5/'XG+"
+
+
+def analyze_func(func):
+    assert inspect.isroutine(func)
+
+    analyzer = CodeAnalyzer()
+    analyzer.analyze(func)
+    return analyzer.get_analyzed_bytes()
+
+
+class CodeAnalyzer:
+    """
+    This class contains the logic to analyze a code object. For now, it
+    only analyzes the function's bytecode and constants. But in the
+    future, it will be able to analyze its references, including class
+    types.
+    """
+
+    def __init__(self):
+        self._id_to_hash = {}
+        self._ids_being_hashed = set()
+        self._analyzed_bytearray = bytearray()
+
+    def get_analyzed_bytes(self):
+        return self._analyzed_bytearray
+
+    def analyze(self, obj):
+        obj_id = get_obj_id(obj)
+
+        # If already hashed, return the hash.
+        if obj_id in self._id_to_hash:
+            return self._id_to_hash[obj_id]
+
+        # If the obj is already hashing, break the circular ref and return early.
+        if obj_id in self._ids_being_hashed:
+            return get_obj_name(obj, DEFAULT_NAME)
+        self._ids_being_hashed.add(obj_id)
+
+        # Otherwise, hash the object.
+        obj_hash = self._analyze(obj)
+        self._id_to_hash[obj_id] = obj_hash
+        self._ids_being_hashed.remove(obj_id)
+
+    def _update_bytes(self, data):
+        self._analyzed_bytearray += data
+
+    def _analyze(self, obj):
+        if isinstance(obj, bytes) or isinstance(obj, bytearray):
+            self._update_bytes(obj)
+
+        elif obj is None or isinstance(obj, (int, float, str, bool)):
+            json_bytes = json.dumps(obj).encode()
+            self._update_bytes(json_bytes)
+
+        elif isinstance(obj, (list, tuple)):
+            for elem in obj:
+                self.analyze(elem)
+
+        elif isinstance(obj, dict):
+            for key, elem in obj.items():
+                self.analyze(key)
+                self.analyze(elem)
+
+        elif inspect.isroutine(obj):
+            if obj.__defaults__ is not None:
+                for d in obj.__defaults__:
+                    self.analyze(d)
+
+            self._analyze_code(obj.__code__)
+
+        elif inspect.iscode(obj):
+            self._analyze_code(obj)
+
+        elif inspect.isclass(obj):
+            # TODO: Add support to analyze class objects.
+            self.analyze(obj.__name__)
+
+        else:
+            message = oneline(
+                f"""
+                Found complex object {obj!r} of type {type(obj)!r} while analyzing
+                code for caching, probably as a global variable or a function
```

Comment:
Do we check globals and defaults right now?

Reply:
Ah, we don't. I have that on another branch and I probably didn't vet this warning before breaking this change from that branch. I'll remove the "probably as a global variable or a function default" bit and add it back when we start looking at them.